File: js/QuickSort.js
class QuickSort {
  constructor() {
  }
  // weird stuff going here
  async Sort(low, high) {
    if (low < high) {
      let pi = await this.partition(low, high);

      await this.Sort(low, pi - 1);
      await this.Sort(pi + 1, high);
    }
  }

  async partition(low, high) {
    let pivot = v[high];
    let i = (low - 1);
    for (let j = low; j < high; j++) {

      if (v[j] <= pivot) {

        for (let l = i + 1; l < j; l++)
          status[l] = 3;

        status[i] = 1;
        status[j] = 2;
        i++;
        
        await this.swap(i, j);

        for (let l = i; l < j; l++)
          status[l] = 0;
        status[i - 1] = 0;
        status[j] = 0;
      }

    }

    let temp = v[i + 1];
    v[i + 1] = v[high];
    v[high] = temp;

    return i + 1;
  }

  async swap(i, j) {
    await sleep();

    let aux = v[i];
    v[i] = v[j];
    v[j] = aux;
  }

}

File: src/main/java/windows/main/sorting/algorithms/QuickSort.java
package windows.main.sorting.algorithms;

import processing.core.PApplet;
import windows.main.sorting.SortingAlgorithm;
import windows.main.sorting.colors.Color;
import windows.main.sorting.colors.Colors;

public class QuickSort extends SortingAlgorithm {

    public QuickSort(PApplet pApplet, int[] vector, Color[] color) {
        super(pApplet, vector, color, "Quick Sort");
    }

    @Override
    public void run() {
        onAlgorithmStart();

        quickSort(0, vector.length - 1);

        onAlgorithmStops();
    }

    void quickSort(int low, int high) {
        comparisons++;
        if (low < high) {
            int pi = partition(low, high);

            if (stop) {
                return;
            }

            setColor(pi, Colors.PIVOT_COLOR);

            quickSort(low, pi - 1);
            quickSort(pi + 1, high);

            resetColor(pi);
        }
    }

    int partition(int low, int high) {
        int pivot = vector[high];
        arrayAccess++;

        int i = (low - 1);

        for(int j = low; j <= high - 1; j++) {
            if (stop) {
                return 0;
            }

            setColor(j, Colors.CURRENT_INDEX);

            comparisons++;
            arrayAccess++;
            if (vector[j] < pivot) {

                i++;

                setColor(i, Colors.SWAPPING_INDEX);

                playSound(i);

                sleep();
                stopSound();

                while (pause) {
                    if (stop) {
                        return 0;
                    }
                    pauseSleep();
                }

                swap(i, j);
                resetColor(i);
            }

            resetColor(j);
        }

        swap(i + 1, high);
        return (i + 1);
    }
}
